File: push_app/ios/OneSignalNotificationServiceExtension/NotificationService.m
#import <OneSignal/OneSignal.h>

#import "NotificationService.h"

@interface NotificationService ()

@property (nonatomic, strong) void (^contentHandler)(UNNotificationContent *contentToDeliver);
@property (nonatomic, strong) UNNotificationRequest *receivedRequest;
@property (nonatomic, strong) UNMutableNotificationContent *bestAttemptContent;

@end

@implementation NotificationService



- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    self.receivedRequest = request;
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];
    
    //If your SDK version is < 3.5.0 uncomment and use this code:
    /*
    [OneSignal didReceiveNotificationExtensionRequest:self.receivedRequest
                       withMutableNotificationContent:self.bestAttemptContent];
    self.contentHandler(self.bestAttemptContent);
    */


  
    
    NSUserDefaults *userDefault = [[NSUserDefaults alloc] initWithSuiteName:@"group.com.kechaou.pushApp.onesignal"];
    NSLog(@"#########  NSE player_id: %@", [userDefault  stringForKey:@"GT_PLAYER_ID"]);
    NSLog(@"######### NSE app_id: %@", [userDefault  stringForKey:@"GT_APP_ID"]);
    
    /* DEBUGGING: Uncomment the 2 lines below and comment out the one above to ensure this extension is excuting
                  Note, this extension only runs when mutable-content is set
                  Setting an attachment or action buttons automatically adds this */


    
 
//    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
//    [userDefaults setObject:@"Dry" forKey:@"flutter.test_key"];
//    [userDefaults setObject:@"Dry" forKey:@"test_key"];
//    [userDefaults synchronize];
    
 

    
    NSLog(@"################1 = %@", self.bestAttemptContent.title);
    NSLog(@"################2 = %@", self.bestAttemptContent.userInfo);
    NSLog(@"################3 = %@", self.bestAttemptContent.attachments);
    
    
    // In body data for the 'application/x-www-form-urlencoded' content type,
    // form fields are separated by an ampersand. Note the absence of a
    // leading ampersand.
    NSString *bodyData = @"name=Jane+Doe&address=123+Main+St";

    NSMutableURLRequest *postRequest = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"https://www.apple.com"]];

    // Set the request's content type to application/x-www-form-urlencoded
    [postRequest setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];

    // Designate the request a POST request and specify its body data
    [postRequest setHTTPMethod:@"POST"];
    [postRequest setHTTPBody:[NSData dataWithBytes:[bodyData UTF8String] length:strlen([bodyData UTF8String])]];

    // Initialize the NSURLConnection and proceed as described in
    // Retrieving the Contents of a URLÏ

    // Create url connection and fire request
    NSURLConnection *conn = [[NSURLConnection alloc] initWithRequest:request delegate:self];

    

    
    [OneSignal setLogLevel:ONE_S_LL_VERBOSE visualLevel:ONE_S_LL_NONE];
   
    [OneSignal didReceiveNotificationExtensionRequest:self.receivedRequest
                       withMutableNotificationContent:self.bestAttemptContent
                                   withContentHandler:self.contentHandler];
}

- (void)serviceExtensionTimeWillExpire {
    // Called just before the extension will be terminated by the system.
    // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
    
    [OneSignal serviceExtensionTimeWillExpireRequest:self.receivedRequest withMutableNotificationContent:self.bestAttemptContent];
    
    self.contentHandler(self.bestAttemptContent);
}

@end

File: push_app/ios/OneSignalNotificationServiceExtension/NotificationService.swift
import UserNotifications
import OneSignal
import os.log

class NotificationService: UNNotificationServiceExtension {
    
    var contentHandler: ((UNNotificationContent) -> Void)?
    var receivedRequest: UNNotificationRequest!
    var bestAttemptContent: UNMutableNotificationContent?
    
    
    override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
        self.receivedRequest = request;
        self.contentHandler = contentHandler
        self.bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
        
        
        
        let userDefaults = UserDefaults(suiteName: "group.com.kechaou.pushApp.onesignal")
        let player_id = userDefaults!.string(forKey: "GT_PLAYER_ID")
        NSLog("################ player_id \(player_id ?? "empty")" )
        
        
        let userInfo = request.content.userInfo
        
        let custom = userInfo["custom"] as? [String: Any]
        let arrAPS = userInfo["aps"] as? [String: Any]
        let arrAlert = arrAPS!["alert"] as? [String:Any]
        
        let not_id:String = custom!["i"] as? String ?? ""
        let strTitle:String = arrAlert!["title"] as? String ?? ""
        let strBody:String = arrAlert!["body"] as? String ?? ""
        
//        NSLog("###### Running NotificationServiceExtension: userInfo.description = \(userInfo.description)")
//        NSLog("####### Running NotificationServiceExtension: custom = \(custom.debugDescription)")
        let received_time = Date()
        let received_time_formatter = DateFormatter()
        received_time_formatter.dateFormat = "y-MM-dd H:mm:ss.SSSS"
        
        NSLog("####### Running NotificationServiceExtension: received_time = \(received_time_formatter.string(from: received_time))")
        NSLog("####### Running NotificationServiceExtension: not_id = \(not_id)")
        NSLog("####### Running NotificationServiceExtension: strTitle = \(strTitle)")
        NSLog("####### Running NotificationServiceExtension: strBody = \(strBody)")

        
        NSLog("###### send req")
        // Create URL
        let url = URL(string: "https://push-notification-admin-panel.herokuapp.com/api/receivedNotification?isIOS=true")
        guard let requestUrl = url else { fatalError() }

        // Create URL Request
        var request_post = URLRequest(url: requestUrl)

        // Specify HTTP Method to use
        request_post.httpMethod = "POST"
        request_post.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: AnyHashable] = [
            "user_id" :player_id,
            "title": strTitle,
            "content": strBody,
            "received_time": received_time_formatter.string(from: received_time),
            "notification_id": not_id
        ]
        
        request_post.httpBody = try? JSONSerialization.data(withJSONObject: body, options: .fragmentsAllowed)
       

        // Send HTTP Request
        let task = URLSession.shared.dataTask(with: request_post) { (data, response, error) in
            
            // Check if Error took place
            if let error = error {
                NSLog("######## Error took place \(error)")
                return
            }
            
            // Read HTTP Response Status code
            if let response = response as? HTTPURLResponse {
                NSLog("######### Response HTTP Status code: \(response.statusCode)")
            }
            
            // Convert HTTP Response Data to a simple String
            if let data = data, let dataString = String(data: data, encoding: .utf8) {
                NSLog("########### Response data string:\n \(dataString)")
            }
            
        }
        task.resume()
        
        

      //debug log types need to be enabled in Console > Action > Include Debug Messages
        os_log("%{public}@", log: OSLog(subsystem: "com.kechaou.pushApp", category: "OneSignalNotificationServiceExtension"), type: OSLogType.debug, userInfo.debugDescription)
        
        if let bestAttemptContent = bestAttemptContent {
            OneSignal.setLogLevel(.LL_VERBOSE, visualLevel: .LL_NONE)
            
            //bestAttemptContent.body = "[Modified2] " + bestAttemptContent.body
            OneSignal.didReceiveNotificationExtensionRequest(self.receivedRequest, with: self.bestAttemptContent, withContentHandler: self.contentHandler)
        }
    }
    
    override func serviceExtensionTimeWillExpire() {
        // Called just before the extension will be terminated by the system.
        // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
        if let contentHandler = contentHandler, let bestAttemptContent =  bestAttemptContent {
            OneSignal.serviceExtensionTimeWillExpireRequest(self.receivedRequest, with: self.bestAttemptContent)
            contentHandler(bestAttemptContent)
        }
    }
}
